### PR TITLE
List/Watch on field=metadata.name= across namespaces is broken

### LINKED ISSUE #13743: Can't watch or list objects with a given name across all namespaces
`generic/etcd` assumes MatchesSingle is that "metadata.name" is set, but if namespace is not set, matches single should be false.  It should be possible to call `GET /api/v1/pods?fields=metadata%2Dname`

One possible fix is to assume that an error in List and Watch in generic/etcd on KeyFunc means that we should not continue down the match single path.

@wojtek-t @lavalamp

### PR DESCRIPTION
**Pull request by smarterclayton**
MatchSingle assumes that namespace is set.  That is not correct.

Fixes #13743

### CODE REVIEW

File: pkg/registry/generic/etcd/etcd.go [hunk 1]
```diff
@@ -152,23 +152,24 @@ func (e *Etcd) ListPredicate(ctx api.Context, m generic.Matcher) (runtime.Object
...
-		if err != nil {
-			return nil, err
-		}
-	} else {
-		trace.Step("About to list directory")
-		err := e.Storage.List(e.KeyRootFunc(ctx), list)
-		trace.Step("List extracted")
-		if err != nil {
-			return nil, err
+		if key, err := e.KeyFunc(ctx, name); err == nil {
```
Comment (lavalamp) [inline pkg/registry/generic/etcd/etcd.go]: I think we should build these smarts into the KeyFunc-- won't this change prevent one from (efficiently) watching an object before it is created?

Comment (smarterclayton) [inline pkg/registry/generic/etcd/etcd.go]: Shouldn't - you have to say what namespace you want to watch in.  The problem is the predictate (metadata.name) does not include the implicit predicate coming off the url (the namespace value).  KeyFunc requires name && namespace > 0.  The only thing this prevents is us rejecting metadata.name= at this level (but it should be higher anyway)

**Comment by wojtek-t** [comment]
LGTM

File: pkg/registry/generic/etcd/etcd.go (hunk 1 above)
Comment (lavalamp) [inline pkg/registry/generic/etcd/etcd.go]: OK.

**Comment by lavalamp** [comment]
LGTM

**Comment by smarterclayton** [comment]
ok to test

**Comment by smarterclayton** [comment]
Green, reading lava lamps LGTM unless he squawks about it...

**Pull request merged**